### PR TITLE
Prepare Content Packs Metadata for Upload

### PR DESCRIPTION
**Pull request by yaakovpraisler**
## Status
- [x] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-4849

## Description
Moved the upload-flow logics of formating the packs metadata to the SDK prepare-content command.

### CODE REVIEW

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 1]
```diff
@@ -194,6 +214,14 @@ def depends_on(self) -> List["RelationshipData"]:
             if r.content_item_to.database_id == r.target_id
         ]
 
+    @property
+    def dependent_packs(self):
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/objects/pack.py:218]: All depends_on of packs are packs, so this can be removed

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 2]
```diff
@@ -303,3 +342,187 @@ def to_nodes(self) -> Nodes:
             self.to_dict(),
             *[content_item.to_dict() for content_item in self.content_items],
         )
+
+    def enhance_pack_properties(self, marketplace: MarketplaceVersions):
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/objects/pack.py:346]: I think we should make a validator for it to enhance those properties automatically.

File: demisto_sdk/commands/content_graph/common.py [hunk 3]
```diff
@@ -108,6 +108,60 @@ def server_name(self) -> str:
...
+            return "Modeling Rule"
+        elif self == ContentType.CORRELATION_RULE:
+            return "Correlation Rule"
+        elif self == ContentType.XSIAM_DASHBOARD:
+            return "XSIAM Dashboard"
+        elif self == ContentType.XSIAM_REPORT:
+            return "XSIAM Report"
+        elif self == ContentType.XDRC_TEMPLATE:
+            return "XDRC Template"
+        return self
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/common.py:163]:
```suggestion
    def metadata_display_name(self) -> str:
        return re.sub(r'([a-z](?=[A-Z])|[A-Z](?=[A-Z][a-z]))', r'\1 ', self.value)
```
note that it won't cover special cases like "Automation" and "Reputation" - is it necessary tho?

File: demisto_sdk/commands/content_graph/common.py [hunk 4]
```diff
@@ -108,6 +108,60 @@ def server_name(self) -> str:
             return "classifier-mapper"
         return self.lower()
 
+    @property
+    def metadata_name(self) -> str:
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/common.py:112]: can we somehow adjust `server_name()`? they are very similar

File: demisto_sdk/commands/content_graph/common.py (hunk 3 above)
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/common.py:163]: Yes. but these cases can be excluded

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 5]
```diff
@@ -233,6 +251,10 @@ def validate_path(cls, v: Path) -> Path:
             return v
         return CONTENT_PATH / v
 
+    @property
+    def is_private(self) -> Optional[bool]:
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/pack.py:255]: sounds like it's not optional

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 6]
```diff
@@ -295,60 +317,32 @@ def set_content_items(self):
 
     def dump_metadata(self, path: Path, marketplace: MarketplaceVersions) -> None:
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/pack.py:318]: there's a lot of logic here which is a bit hard to follow - please add a docstring with a detailed description, I will go over this afterward

**Review by dantavori — CHANGES_REQUESTED** [review]
very nice - in my review I didn't compare the fields with the current implementation in content, leaving this to you by using the comparison between the buckets before and after your changes, as discussed.

File: demisto_sdk/commands/content_graph/objects/pack.py
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/pack.py:1]: to avoid a too-large file that isn't readable, let's split the methods in `Pack` class into two:
1. the "core" methods will remain here (properties, getters/setters)
2. the rest of the methods should be related to upload. move them to a class `UploadablePack` (in a different file). You may need to define some abstract methods/properties, and consider also just moving properties that are related only to the upload to it.
Then make Pack inherit UploadablePack.

In addition, let's add a `_` prefix to methods that shouldn't be "exposed" not from within the class's inner methods.

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 7]
```diff
@@ -603,3 +603,368 @@ def to_nodes(self) -> Nodes:
...
+        """
+        return {
+            r.content_item_to.object_id: {
+                "mandatory": r.mandatorily,
+                "minVersion": r.content_item_to.current_version,  # type:ignore[attr-defined]
+                "author": r.content_item_to.author  # type:ignore[attr-defined]
+                if marketplace == MarketplaceVersions.XSOAR
+                else r.content_item_to.author.replace(  # type:ignore[attr-defined]
+                    "XSOAR", "XSIAM"
+                ),
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/pack.py:736]: let's use the same method I suggested above

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 8]
```diff
@@ -603,3 +603,368 @@ def to_nodes(self) -> Nodes:
...
+        item_type_key (Optional[str]): The content item type key to search in its list value that exists in the collected_content_items dict.
+
+    Returns:
+        list: List of the found content items.
+    """
+    return [
+        content_item
+        for content_item in collected_content_items[item_type_key]
+        if content_item.get("id") == item_id and content_item.get("name") == item_name
+    ]
```
[the remark this answers is not in the document]
Comment (dantavori) [reply demisto_sdk/commands/content_graph/objects/pack.py:970]: just a sanity check, are we sure that in this stage, name, id, and type represent a unique content item? it's not possible here to have two versions of the same item but with different fromversion/toversion fields?

**Review by dantavori — CHANGES_REQUESTED** [review]
looks good, we are getting there!

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 9]
```diff
@@ -603,3 +603,368 @@ def to_nodes(self) -> Nodes:
...
+
+        Args:
+            marketplace (MarketplaceVersions): The marketplace to which the pack should belong to.
+        """
+        self.tags = self.get_pack_tags(marketplace)
+        self.author = (
+            self.author
+            if marketplace == MarketplaceVersions.XSOAR
+            else self.author.replace("XSOAR", "XSIAM")  # type:ignore[union-attr]
+        )
```
[the remark this answers is not in the document]
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/objects/pack.py:623]: done

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 10]
```diff
@@ -603,3 +603,368 @@ def to_nodes(self) -> Nodes:
...
+            or str(
+                max(
+                    (
+                        parse(content_item.fromversion)
+                        for content_item in self.content_items
+                    ),
+                    default=MARKETPLACE_MIN_VERSION,
+                )
+            )
+            or MARKETPLACE_MIN_VERSION
```
[the remark this answers is not in the document]
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/objects/pack.py:636]: removed

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 11]
```diff
@@ -603,3 +603,368 @@ def to_nodes(self) -> Nodes:
...
+        )
+
+    def enhance_metadata(self, marketplace: MarketplaceVersions) -> dict:
+        """
+        Enhancing the pack metadata properties after dumping into a dictionary. (properties that can't be calculating before)
```
[the remark this answers is not in the document]
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/objects/pack.py:641]: because I prefer not to change the object property value just for the metadata, for example, self.content_items is calculated by the graph/pydantic etc., so I don't think we need to change it because in the metadata it should have a different value

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 12]
```diff
@@ -603,3 +603,368 @@ def to_nodes(self) -> Nodes:
...
+            marketplace (MarketplaceVersions): The marketplace to which the pack should belong to.
+
+        Returns:
+            Tuple[Dict, Dict]: The content items and display names dictionaries to add to the pack metadata.
+        """
+        content_items: dict = {}
+        content_displays: dict = {}
+        for content_item in self.content_items:
+
+            if content_item.content_type == ContentType.TEST_PLAYBOOK:
```
[the remark this answers is not in the document]
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/objects/pack.py:691]: done

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 13]
```diff
@@ -603,3 +603,368 @@ def to_nodes(self) -> Nodes:
...
+                logger.debug(
+                    f"Skip loading the {content_item.name} test playbook into metadata.json"
```
[the remark this answers is not in the document]
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/objects/pack.py:693]: done

File: demisto_sdk/commands/content_graph/objects/pack.py (hunk 7 above)
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/objects/pack.py:736]: done

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 14]
```diff
@@ -603,3 +603,368 @@ def to_nodes(self) -> Nodes:
...
+        content_item_summary (dict): The current content item summary to update if needed.
+    """
+    if parse(content_item.toversion) > parse(
+        content_item_metadata["toversion"] or DEFAULT_CONTENT_ITEM_TO_VERSION
+    ):
+        logger.debug(
+            f'Current content item with name "{content_item.name}" has higher `toversion` than the existing object, '
+            "updating its metadata."
+        )
+        content_item_metadata.update(content_item_summary.items())
```
[the remark this answers is not in the document]
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/objects/pack.py:930]: done

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 15]
```diff
@@ -603,3 +603,368 @@ def to_nodes(self) -> Nodes:
...
+        set_empty_toversion_if_default(content_item_metadata)
```
[the remark this answers is not in the document]
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/objects/pack.py:931]: don't think we need since most of the items has the fromversion value

File: demisto_sdk/commands/content_graph/objects/pack.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/objects/pack.py:970]: done.

Yes it's possible, and that's why we are filtering the item with the lowest version here so only the item with the higher version will appear in the metadata.

File: demisto_sdk/commands/content_graph/objects/pack.py
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/objects/pack.py:1]: done

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 16]
```diff
@@ -288,67 +175,56 @@ def set_content_items(self):
...
-                min(parse(content_item.fromversion) for content_item in content_items)
+                min(
+                    parse(content_item.fromversion)
+                    for content_item in content_items
+                    if content_item.content_type
+                    not in (
+                        ContentType.XDRC_TEMPLATE,
+                        ContentType.TRIGGER,
+                        ContentType.TEST_PLAYBOOK,
+                    )
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/pack.py:186]: feels risky, how can we ensure it's maintained? e.g. we soon [might add](https://jira-hq.paloaltonetworks.local/browse/CRTX-89006) a fromversion field to triggers.

File: demisto_sdk/commands/content_graph/objects/pack_metadata.py [hunk 17]
```diff
@@ -0,0 +1,477 @@
...
+        """
+        support_details = {}
+        if self.url:
+            support_details["url"] = self.url
+        if self.email:
+            support_details["email"] = self.email
+        return support_details
+
+
+def get_author(author, marketplace):
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/pack_metadata.py:349]: please move the below methods to the PackMetadata class scope (as they shouldn't be used outside of this scope).

File: demisto_sdk/commands/content_graph/parsers/pack.py [hunk 18]
```diff
@@ -143,7 +143,10 @@ def __init__(self, path: Path, metadata: Dict[str, Any]) -> None:
         self.commit: str = GitUtil().get_current_commit_hash() or ""
         self.downloads: int = 0
         self.tags: List[str] = metadata["tags"] or []
-        self.categories: List[str] = [c.title() for c in metadata["categories"]]
+        self.categories: List[str] = [
+            " ".join([w.title() if w.islower() else w for w in c.split()])
+            for c in metadata["categories"]
+        ]
```
[the remark this answers is not in the document]
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/parsers/pack.py:138]: added as `@property`

File: demisto_sdk/commands/content_graph/parsers/pack.py [hunk 19]
```diff
@@ -157,11 +160,14 @@ def __init__(self, path: Path, metadata: Dict[str, Any]) -> None:
...
-        self.preview_only: Optional[bool] = metadata.get("previewOnly")
-        self.disable_monthly: Optional[bool] = metadata.get("disableMonthly")
+        self.vendor_id: Optional[str] = metadata.get("vendorId") or ""
+        self.partner_id: Optional[str] = metadata.get("partnerId") or ""
+        self.partner_name: Optional[str] = metadata.get("partnerName") or ""
+        self.preview_only: Optional[bool] = metadata.get("previewOnly") or False
+        self.disable_monthly: Optional[bool] = metadata.get("disableMonthly") or False
+        self.content_commit_hash: Optional[str] = (
+            metadata.get("contentCommitHash") or ""
+        )
```
[the remark this answers is not in the document]
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/parsers/pack.py:159]: I'm not sure... this is what we are doing now in `marketplaces_services`

File: demisto_sdk/commands/content_graph/objects/pack.py (hunk 16 above)
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/objects/pack.py:186]: We can maybe add instead:
```
it content_item.fromversion != "0.0.0"
```

Comment (dantavori) [reply demisto_sdk/commands/content_graph/objects/pack.py:186]: ok, use `DEFAULT_CONTENT_ITEM_FROM_VERSION`

File: demisto_sdk/commands/content_graph/parsers/pack.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (dantavori) [reply demisto_sdk/commands/content_graph/parsers/pack.py:158]: let's talk about it

**Review by dantavori — APPROVED** [review]

File: demisto_sdk/commands/content_graph/tests/parsers_and_models_test.py [hunk 20]
```diff
@@ -1517,15 +1525,18 @@ def test_pack_parser(self, repo: Repo):
...
             expected_author="Cortex XSOAR",
-            expected_certification="",
+            expected_certification="verified",
             expected_hidden=False,
             expected_current_version="1.2.12",
-            expected_tags=[],
+            expected_tags=["TIM"],
             expected_categories=["Utilities"],
-            expected_use_cases=[],
+            expected_use_cases=["Identity And Access Management"],
```
[the remark this answers is not in the document]
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/tests/parsers_and_models_test.py:1539]: Yes

File: demisto_sdk/commands/content_graph/objects/integration.py [hunk 21]
```diff
@@ -70,15 +71,30 @@ def set_commands(self):
         ]
         self.commands = commands
 
+    def summary(
+        self,
+        marketplace: Optional[MarketplaceVersions] = None,
+        incident_to_alert: bool = False,
+    ) -> dict:
+        summary = super().summary(marketplace, incident_to_alert)
+        summary["name"] = self.unified_data.get("display")
```
[the remark this answers is not in the document]
Comment (yaakovpraisler) [reply demisto_sdk/commands/content_graph/objects/integration.py:80]: There is already an attribute `display` but we need the value of `display` with the key `name`, but we have another key `name` with the value of `name`.
